Clasrel

Silvicek opened the pull request:
Adds Argus yes/no task and data set

Silvicek committed:
yes/no task ready

Silvicek committed:
argus yes/no data set

Silvicek committed:
argus yes/no data set update

Silvicek committed:
'task>model' prioritizes task over model config

Silvicek committed:
revert some functionality from argus sts version

Silvicek committed:
same params as argus

Silvicek committed:
fix predict

Silvicek committed:
fix eval

Silvicek committed:
fix settings propagation

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+
```
Interesting.  I have expected that you'd just modify the hypev task appropriately.  Is there a reason to keep the old hypev task in addition to the yesno task?  Is "yesno" a fitting name?  What about the old hypev argus data?

pasky replied:
Actually, it might make sense to have two models that are similar, but one uses the hypothesis aggregation like yesno while the other just averages pair predictions, for comparison purposes.  But they must at least be able to process the same dataset, so could we use the hypev format and load_hypev()?

Silvicek replied:
The hypev task learns on separate samples and the averaging is done only when evaluating. Load_hypev is therefore unusable in yesno. I completely forgot about hypev and will delete it..
The old argus data are still usable in yesno (but should we use it?)

pasky replied:
On Sun, Apr 24, 2016 at 02:23:57AM -0700, Silvicek wrote:

> > @@ -0,0 +1,301 @@
> > +
> 
> The hypev task learns on separate samples and the averaging is done only when evaluating. Load_hypev is therefore unusable in yesno. In my opinion the hypev task is outdated..

I don't understand why does the averaging imply that load_hypev() is
unusable, can you elaborate?

What we need to show in any papers we publish is that our method has
merit compared to a simpler averaging baseline.

Silvicek replied:
load_hypev is unusable because the training only works with pairs <question>-<sentence> and not with <question>-<bunch of sentences>. The precision hypev shows is only averaged over questions after the training. 
The correct averaging baseline should use the same end-to-end approach as the current clasrel implementation only with simple average at the end.

pasky replied:
First, I'm not sure what does that have to do with loading data.
load_anssel doesn't aggregate the second sentences either, does it?

Second, if "hypev" is obsolete, why not name this "hypev"?  I'm not sure
how appropriate "yesno" is e.g. for entrance exam datasets.  It's just
nitpicking, of course, but I don't understand the name change.

Silvicek replied:
Sure i'll rename it :P

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+
+
+class Container:
+    """Container for merging questions together."""
```
Isn't it about merging atexts (etexts) for the same question, rather than merging questions?

Silvicek replied:
yup :(

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+        model = build_model(self.emb, self.vocab, module_prep_model, self.c)
+
+        # for lname in self.c['fix_layers']:  # whats this??
+        #     model.nodes[lname].trainable = False
```
What's the reason to comment this out and not to support fixing some layers in place?

Silvicek replied:
wasnt sure what this does, will uncomment

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+               % (pfx, mres[self.trainf]['Loss'],
+                  pfx, mres[self.valf]['Precision'],
+                  pfx, mres[self.testf].get('Loss', np.nan),
+                  pfx, mres[self.testf].get('Precision', np.nan)))
```
Interesting choice.  Why train + test reports loss (what does it tell us compared to precision?) while val reports precision?

Silvicek replied:
my bad

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+
+    return N
+
+def build_model(glove, vocab, module_prep_model, c):
```
Any reason why this isn't a method of the task?

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+def embedding(model, glove, vocab, s0pad, s1pad, dropout, trainable=False,
+              add_flags=False):
+    """ Sts embedding layer, without creating inputs. """
+    # TODO: add switch to original -> no code duplication
```
Please do. :)

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+    return model
+
+def mlp_ptscorer(model, inputs, Ddim, N, l2reg, pfx='out', sum_mode='sum'):
+    """ Element-wise features from the pair fed to an MLP. """
```
Please make ptscorer configurable.  Can we make normal mlp_ptscorer work like this in case Ddim == 0?

pasky commented on tasks/yesno.py:
```diff
@@ -0,0 +1,301 @@
+    return pfx+'mlp'
+
+from models.pysts.kerasts.blocks import LambdaMerge
+def absdiff_merge(model, layers):
```
Why do we need own copy of this?

pasky commented on pysts/clasrel_layers.py:
```diff
@@ -0,0 +1,133 @@
+from keras.layers.wrappers import TimeDistributed as td
```
Should this file be in pysts/ instead of kerasts/ ?

pasky commented on pysts/clasrel_layers.py:
```diff
@@ -0,0 +1,133 @@
+import numpy as np
+
+def relu(x):
+    return K.switch(x > 0, x + 0.01, 0.01)
```
Huh, keras doesn't have relu, or what's going on?

Silvicek replied:
Before the introduction of stabilizing constant in the clasrel denominator, this was a necessary relu adjustment, now its outdated.

pasky commented:
Thanks! It's great work and I'm very happy that it's finally coming back to dataset-sts.

I think it needs a little bit more work though, I have expected this would share more with the hypev task; at the very least, it should share the data format.  Maybe we can rename the tasks to have more similarity in the name too.

I also don't really like the copied over functions, imho the task structure is slightly messy :) - more in the comments.  I'll be also happy if you could add some more comments, esp. descriptions of modules and classes, but this is lower priority.

Silvicek committed:
fixed masking of max_sentences, simplified layers

Silvicek committed:
dealing with some pull request comments

Silvicek committed:
most pull request notes addressed

Silvicek committed:
renaming stuff

Silvicek committed:
argus data anssel->hypev

Silvicek committed:
added docs

Silvicek committed:
additional ptscorer created, the new mlp led to performance decrease

Silvicek committed:
revert original mlp_ptscorer

Silvicek committed:
fix wrong property

Silvicek committed:
fix layer identifier

Silvicek committed:
no extra inputs

Silvicek commented:
Ok, i addressed most of the issues. It was necessary to create additional ptscorer because the use of the hidden layer in mlp_ptscorer caused val performance decrease.

pasky commented on pysts/vocab.py:
```diff
@@ -56,7 +56,12 @@ def embmatrix(self, emb):
+                else:
+                    embedding_weights[index, :] = np.random.uniform(-0.25, 0.25, emb.N)  # 0.25 is embedding SD
+
+
```
Adventurous change!  However, it'll impact all other models too.  I have started a sanity check evaluation on yoda dataset with just this change to confirm if we'll need to rebenchmark everything again or it's just a small change.

Silvicek replied:
Come to think about it, because of the implemented masking, reverting this should not make any impact on the clasrel results. However using nonzero vectors for nonexistent words is very weird.

pasky replied:
It looks clear to change, no impact. I agree it's nicer.

pasky commented on pysts/kerasts/blocks.py:
```diff
@@ -185,6 +186,20 @@ def mlp_ptscorer(model, inputs, Ddim, N, l2reg, pfx='out', sum_mode='sum', extra
 
+def mulsum_ptscorer(model, inputs, Ddim, N, l2reg, pfx='out', sum_mode='sum'):
+    """ Element-wise features from the pair fed to a concat of sum and multiplication of inputs.
+    """
```
Hmm, I still don't quite understand why we can't just special-case mlp_ptscorer for Ddim=0.   But we don't need to dwell on this, I can try to change that myself.

pasky commented on tasks/hypev.py:
```diff
@@ -3,95 +3,249 @@
+        return('%s%.6f |%s%.6f |%s%.6f '
+               % (pfx, mres[self.trainf]['Precision'],
+                  pfx, mres[self.valf]['Precision'],
+                  pfx, mres[self.testf].get('Precision', np.nan)))
```
So, is it precision or accuracy? I thought .evaluate() reports accuracy, and that's probably what we want too.

pasky commented:
(I think we are clear to merge, except I'd love to clear up the Precision - QAccuracy change logic.)

pasky committed:
tasks/hypev: yesno -> hypev remnants

pasky committed:
tasks/hypev: Use eval_hypev() again, report Accuracy and F1

pasky committed:
tasks/hypev: Select best training epoch based on accuracy, not loss

pasky commented:
thx
